### PR TITLE
defer: inline fragment warning

### LINKED ISSUE #3158: Maybe: Add warning that deferred inline fragments without type condition are not supported
The following is a valid deferred inline fragment but [Apollo iOS will not support it](https://apollograph.slack.com/archives/C054R4FEW7L/p1690479665570519). We should add a warning to the JS frontend when we detect this type of deferred fragment.
```
query Test {
  ... @defer {
    name
  }
}
```

### PR DESCRIPTION
**Pull request by BobaFetters**
-Adding validation error to graphql-js for deferred inline fragments with no type condition

Closes #3158

### CODE REVIEW

**Review by AnthonyMDev — APPROVED** [review]
This looks great! Nice job, and so quick!

File: Sources/ApolloCodegenLib/Frontend/JavaScript/src/__tests__/deferCompilationTests.ts [hunk 1]
```diff
@@ -152,4 +154,46 @@ describe("given schema", () => {
     });
   });
 
+  describe("query has inline fragment with @defer directive and no type condition should throw error", () => {
```
Comment (calvincestari) [inline Sources/ApolloCodegenLib/Frontend/JavaScript/src/__tests__/deferCompilationTests.ts:157]:
```suggestion
  describe("query has inline fragment with @defer directive and no type condition", () => {
```

File: Sources/ApolloCodegenLib/Frontend/JavaScript/src/__tests__/deferCompilationTests.ts [hunk 2]
```diff
@@ -152,4 +154,46 @@ describe("given schema", () => {
...
+        }
+      }
+    }
+    `;
+
+    const document: DocumentNode = parseOperationDocument(
+      new Source(documentString, "Test Query", { line: 1, column: 1 })
+    );
+
+    it("should compile inline fragment with directive", () => {
```
Comment (calvincestari) [inline Sources/ApolloCodegenLib/Frontend/JavaScript/src/__tests__/deferCompilationTests.ts:172]:
```suggestion
    it(" should throw error", () => {
```

File: Sources/ApolloCodegenLib/Frontend/JavaScript/src/__tests__/deferCompilationTests.ts [hunk 3]
```diff
@@ -152,4 +154,46 @@ describe("given schema", () => {
...
+      const validationErrors: readonly GraphQLError[] = validateDocument(schema, document, emptyValidationOptions)
+      expect(validationErrors.length).toEqual(1)
+    });
+  });
+
+  describe("query has inline fragment with @defer directive should not throw error", () => {
```
Comment (calvincestari) [inline Sources/ApolloCodegenLib/Frontend/JavaScript/src/__tests__/deferCompilationTests.ts:178]: I believe this case is already covered by the test up on [line 37](https://github.com/apollographql/apollo-ios/pull/3159/files#diff-9764f88416cc3dd09bf331172384c2bb39d377cf1547465df39e49755aad50c6R37). Can remove this as a duplicate unless I'm missing something here.

File: Sources/ApolloCodegenLib/Frontend/JavaScript/src/validationRules.ts [hunk 4]
```diff
@@ -65,6 +67,25 @@ export function NoTypenameAlias(context: ValidationContext) {
   };
 }
 
+export function DeferredInlineFragmentNoTypeCondition(context: ValidationContext) {
+  return {
+    InlineFragment(node: InlineFragmentNode) {
+      if (node.directives) {
+        for (const directive of node.directives) {
+          if (directive.name.value == "defer" && node.typeCondition == undefined) {
```
Comment (calvincestari) [inline Sources/ApolloCodegenLib/Frontend/JavaScript/src/validationRules.ts:75]: `"defer"` should already be available in `GraphQLDeferDirective.name`. Probably be better to use that than duplicate strings.

**Review by calvincestari — CHANGES_REQUESTED** [review]
I've left a couple suggestions, looks good though.

**Review by calvincestari — APPROVED** [review]

**Comment by calvincestari** [comment]
Thanks @BobaFetters!

**Pull request merged**